nomad/consul.go

```diff
@@ -435,3 +443,107 @@ func (s *Server) purgeSITokenAccessors(accessors []*structs.SITokenAccessor) err
 	_, _, err := s.raftApply(structs.ServiceIdentityAccessorDeregisterRequestType, request)
 	return err
 }
+
+// ConsulConfigsAPI is an abstraction over the consul/api.ConfigEntries API used by
+// Nomad Server.
+//
+// Nomad will only perform write operations on Consul Ingress Gateway Configuration Entries.
+// Removing the entries is not particularly safe, given that multiple Nomad clusters
+// may be writing to the same config entries, which are global in the Consul scope.
+type ConsulConfigsAPI interface {
+	// SetIngressGatewayConfigEntry adds the given ConfigEntry to Consul, overwriting
+	// the previous entry if set.
+	SetIngressGatewayConfigEntry(ctx context.Context, service string, entry *structs.ConsulIngressConfigEntry) error
+
+	// Stop is used to stop additional creations of Configuration Entries. Intended to
+	// be used on Nomad Server shutdown.
+	Stop()
+}
+
+type consulConfigsAPI struct {
+	// configsClient is the API subset of the real Consul client we need for
+	// managing Configuration Entries.
+	configsClient consul.ConfigAPI
+
+	// limiter is used to rate limit requests to Consul
+	limiter *rate.Limiter
+
+	// logger is used to log messages
+	logger hclog.Logger
+
+	// lock protects the stopped flag, which prevents use of the consul configs API
+	// client after shutdown.
+	lock    sync.Mutex
+	stopped bool
+}
+
+func NewConsulConfigsAPI(configsClient consul.ConfigAPI, logger hclog.Logger) *consulConfigsAPI {
+	return &consulConfigsAPI{
+		configsClient: configsClient,
+		limiter:       rate.NewLimiter(configEntriesRequestRateLimit, int(configEntriesRequestRateLimit)),
+		logger:        logger,
+	}
+}
+
+func (c *consulConfigsAPI) Stop() {
+	c.lock.Lock()
+	defer c.lock.Unlock()
+	c.stopped = true
+}
+
+func (c *consulConfigsAPI) SetIngressGatewayConfigEntry(ctx context.Context, service string, entry *structs.ConsulIngressConfigEntry) error {
+	configEntry := convertIngressGatewayConfig(service, entry)
+	return c.setConfigEntry(ctx, configEntry)
+}
+
+// setConfigEntry will set the Configuration Entry of any type Consul supports.
+func (c *consulConfigsAPI) setConfigEntry(ctx context.Context, entry api.ConfigEntry) error {
+	defer metrics.MeasureSince([]string{"nomad", "consul", "create_config_entry"}, time.Now())
+
+	// make sure the background deletion goroutine has not been stopped
+	c.lock.Lock()
+	stopped := c.stopped
+	c.lock.Unlock()
+
+	if stopped {
+		return errors.New("client stopped and may not longer create config entries")
```

Comment:
Is there a risk of this getting emitted at shutdown?

I can try to determine that but need to stop for the day.